### ISSUE TITLE
ImRect::Contains - Inconsistent edge checking

### ISSUE THREAD

**Issue by Mrgoodbytes13** [ocornut/imgui#3976]
**Version/Branch of Dear ImGui:**

Version: 1.83 WIP
Branch: master

**Issue:**

In the ImVec2 version of ImRect.Contains, the upper-left corners are checked with >= and the lower-right are checked with <

**Example:**
```
ImRect rect = ImRect(0, 0, 1, 1);

bool bl = rect.Contains(ImVec2(0, 0));      // true
bool mid = rect.Contains(ImVec2(0.5, 0.5)); // true
bool tr = rect.Contains(ImVec2(1, 1));      // false
```

**Comment by PathogenDavid** [comment]
That's not by accident. It's pretty common to have the left-top coordinate of a rectangle be inclusive and the bottom-right be exclusive. It allows you to more easily represent a rectangle with no area (IE: `(0, 0, 0, 0)`) and to calculate the size by subtracting the opposing dimensions. (Width is `Right - Left`)

If the right-bottom coordinate was inclusive, a rectangle with no area becomes `(0, 0, -1, -1)` and the size requires adding the missing pixel back in. (Width is `Right - Left + 1`)

**Comment by Mrgoodbytes13** [comment]
Interesting... Thanks for the explanation

**Comment by ocornut** [comment]
Similar as #3301 + some of #1615 (but #1615 also compare hit testing with rendering and exact layout of rasterized pixels, which itself is another topic).

I'm not sure I fully follow the part of @PathogenDavid logic about adding +1 in calculating size? 
I feel we could (but I don't know if we should) use <= in the tests and one side-effect is that zero-sized rectangle would indeed catch some points. But I'm not sure I know if this would cause problems. 

Bounding box are either built from 2 points, or starting with (+FLT_MAX,+FLT_MAX,-FLT_MAX,-FLT_MAX) and then using `Add()`.

Someone else sent me some thoughts by e-mail (ref to myself: 2020/10/13)... and loosely translating then here:
- they assume that the reason that on touching ImRect (aligned edges) a given point will only be part of a single ImRect.
- they noticed that ImPlot used it (at least in Oct 2020) to test for adding points in graphs and as a result it was impossible to add a point exactly on the X axis or the right border of the graph.
- they point out very rightfully that it create a inconsistency with using `ImRect::Add(const ImVec2& p)` to add point to a bounding box, and that in particular seems like a problem considering the Add() function exist exactly for this.

That last point alone makes it tempting to make that change to be honest...

**Comment by ocornut** [comment]
I'll reopen this for now as it seems worth thinking about.

**Comment by PathogenDavid** [comment]
> I'm not sure I fully follow the part of @PathogenDavid logic about adding +1 in calculating size?

I guess it depends whether you consider coordinates in Dear ImGui to lie on a pixel grid or not. As a (primarily) 2D graphics library, I'd argue it only makes sense that they do. Anything else is asking for trouble when introducing anything that intrinsically uses whole number pixel coordinates. (Like mouse cursor positions or platform windows/monitors.)

For example, if I have the following rectangle on a pixel grid where all coordinates are whole numbers and I ask "Is my mouse cursor at `(7, 7)` within the rectangle?" I would expect the answer to be "No".

![image](https://user-images.githubusercontent.com/278957/112834603-ccf68b80-905d-11eb-88a3-0ad0400ec8b8.png)

Every discretely and identically sized pixel begins somewhere and ends somewhere else, regardless of how it is rasterized. It makes sense for pixel N to span `[N, N + 1)`. Using inclusive bounds where pixel N spans `[N, N + 1]` means that all pixels share an infinitely thin strip of the coordinate space in each direction with their immediate neighbors.

For anything not dealing in pixels (like rendering the UI in 3D), these kinds of tiny "errors" don't matter. (If they do matter, you already lost by using floats.)

> they point out very rightfully that it create a inconsistency with using ImRect::Add(const ImVec2& p) to add point to a bounding box, and that in particular seems like a problem considering the Add() function exist exactly for this.

This is definitely somewhat problematic. I think the better solution here is to change the math so adding a coordinate outside the rectangle expands the rectangle to actually include the point:

```cpp
void Add(const ImVec2& p)
{
    if (Min.x > p.x)
        Min.x = p.x;

    if (Min.y > p.y)
        Min.y = p.y;
    
    if (Max.x <= p.x)
        Max.x = nextafterf(p.x, FLT_MAX);
    
    if (Max.y <= p.y)
        Max.y = nextafterf(p.y, FLT_MAX);
}
```

(Although I would question if this user is really intending to add infinitely-small points or if they really want to add whole pixels. Because for whole pixels you probably want to store `p.x + 1.f` instead. The only place where this API is used within Dear ImGui is realistically not affected by this issue.)

----------------

I think much of the confusion in this issue and others potentially stems from the name "max" implying that the coordinate is part of the rectangle even though all the comments describe it as a right-bottom edge.

I don't think these users are expecting hit tests to include the infinitely thin lines on the right and bottom edges of the rectangle. They're expecting the `Max` coordinate to represent the right-bottom pixel which is lit to contribute to the visual of the rectangle.

To put it another way, the term "max" (and even "right-bottom") is ambiguous between the blue and green dots in this image:

![image](https://user-images.githubusercontent.com/278957/112836054-b3eeda00-905f-11eb-8fa7-f9af2a8cb2b7.png)

Personally this is why I prefer surfacing `(x, y, width, height)` for 2D APIs. It side-steps this whole issue since people can generally agree on the pixel dimensions of things.
